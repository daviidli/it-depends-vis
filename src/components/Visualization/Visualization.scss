#visualization {
	.implementation {
		stroke-dasharray: 10 4;
	}

	.dependency {
		stroke-dasharray: 10 4;
	}
}