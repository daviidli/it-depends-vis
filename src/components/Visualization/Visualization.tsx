import React, { useEffect } from 'react';
import * as d3 from 'd3';
import GraphData, { INode, IEdge } from '../../data/GraphData';

const Visualization = () => {
	const id = 'visualization';

	const sampleResponse = {
		names: ['constructor (A)', 'main (A)', 'receiveInput (A)', 'constructor (B)', 'doSomething (B)', 'doSomethingElse (B)', 'constructor (C1)', 'foo (C1)', 'constructor (C2)', 'barbar (C2)'],
		size: [1,1,6,2,4,1,1,2,1,1],
		data: [
			[1,1,0.166666667,0.5,0.25,0,0,0,0,0],
			[1,1,1,1,0.75,1,1,1,1,0],
			[1,1,0.333333333,1,0.25,1,0,0,0,0],
			[1,1,0.5,0.5,1,0,0,0.5,0,0],
			[0,0,0.166666667,0.5,0,1,0,0,0,0],
			[0,0,0.166666667,0,0,0,1,0.5,1,0],
			[0,0,0.333333333,0,0.25,0,1,1,1,0],
			[0,0,0.166666667,0,0,0,1,0.5,1,0],
			[0,0,0,0,0,0,0,0,0,1]
		]
	};

	// const sampleResponse = {
	// 	names: ['A', 'B'],
	// 	size: [2, 4],
	// 	data: [
	// 		[1, 0.3],
	// 		[0.6, 1]
	// 	]
	// }

	const graphData: GraphData = new GraphData(sampleResponse);
	const nodes: any[] = graphData.getNodes();
	const edges: any[] = graphData.getEdges();

	const drag = (simulation: any): any => {
		const onDragStart = (d: any) => {
			if (!d3.event.active) simulation.alphaTarget(0.3).restart();
			d.fx = d.x;
			d.fy = d.y;
		}
		
		const onDrag = (d: any) => {
			d.fx = d3.event.x;
			d.fy = d3.event.y;
		}
		
		const onDragEnd = (d: any) => {
			if (!d3.event.active) simulation.alphaTarget(0);
			d.fx = null;
			d.fy = null;
		}
		
		return d3.drag()
			.on('start', onDragStart)
			.on('drag', onDrag)
			.on('end', onDragEnd);
	};

	useEffect(() => {
		const width = window.innerWidth;
		const height = window.innerHeight;	
	
		const svg = d3.select('#' + id)
			.attr('width', width)
			.attr('height', height);

		const simulation = d3.forceSimulation(nodes)
			.force('link', d3.forceLink(edges).id((edge: any) => edge.id).distance(500).strength(0.3))
			.force('charge', d3.forceManyBody().strength((node: any) => node.size * -500))
			.force('center', d3.forceCenter(width / 2, height / 2));

		const link = svg.append('g')
			.selectAll('path')
			.data(edges)
			.enter()
			.insert('path')
				.attr('class', 'link')
				.attr('stroke', '#000')
				.attr('stroke-width', (edge) => Math.sqrt(edge.weight * 10))
				.attr('fill', 'none')
				.attr('marker-end', function(d) { return 'url(#arrow)'; })

		const node = svg.append('g')
				.attr('stroke', '#fff')
				.attr('stroke-width', 1.5)
			.selectAll('circle')
			.data(nodes)
			.join('circle')
				.attr('r', (d: any) => d.size * 10)
				.attr('fill', '#fedcba')
				.call(drag(simulation));

		svg.append('defs').selectAll('marker')
				.data(['arrow'])
			.enter().append('marker')
				.attr('id', (type) => type)
				.attr('viewBox', '0 -5 10 10')
				.attr('refX', 0)
				.attr('refY', 0)
				.attr('markerWidth', 6)
				.attr('markerHeight', 6)
				.attr('orient', 'auto')
			.append('path')
				.attr('d', 'M0,-5L10,0L0,5');

		simulation.on('tick', () => {
			link
				.attr('d', (d) => {
					const dx = d.target.x - d.source.x;
					const dy = d.target.y - d.source.y;
					const dr = Math.sqrt(dx * dx + dy * dy);
					return 'M' + d.source.x + ',' + d.source.y + 'A' + dr + ',' + dr + ' 0 0,1 ' + d.target.x + ',' + d.target.y;
				})
				.attr('d', function(d) {
					const pathLength = this.getTotalLength();
					const markerLength = 5 * Math.sqrt(d.weight * 10);
					const markerSize = Math.sqrt(Math.pow(markerLength, 2) * 2);
					const radius = (d.target.size * 10) + markerSize;

					const m = this.getPointAtLength(pathLength - radius);
					const dx = m.x - d.source.x;
					const dy = m.y - d.source.y;
					const dr = Math.sqrt(dx * dx + dy * dy);

					return 'M' + d.source.x + ',' + d.source.y + 'A' + dr + ',' + dr + ' 0 0,1 ' + m.x + ',' + m.y;
				});
			node
				.attr('cx', (node) => node.x)
				.attr('cy', (node) => node.y);
		});
	}, []);
	
	
	return (
		<svg id={id}></svg>
	);
};

export default Visualization;
