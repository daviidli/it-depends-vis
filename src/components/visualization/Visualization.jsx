/* eslint-disable react/forbid-prop-types */
import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { toast } from 'react-toastify';
import { useHistory } from 'react-router';
import { Spin } from 'antd';
import Header from '../../containers/HeaderContainer';
import D3 from '../../containers/D3Container';
import fetchData from './fetchData';
import routes from '../../constants/routes.json';

const Visualization = ({
	startCommit, endCommit, repo, data, setData
}) => {
	const history = useHistory();
	const [windowSize, setWindowSize] = useState([window.innerWidth, window.innerHeight]);

	useEffect(() => {
		const onError = err => {
			toast.error(err);
			history.push(routes.HOME);
		};

		if (repo.length === 0) {
			history.push(routes.HOME);
		} else {
			fetchData(onError, setData, startCommit, endCommit, repo);
		}
	}, [startCommit, endCommit, setData, repo, history]);

	useEffect(() => {
		const onResize = () => {
			setWindowSize([window.innerWidth, window.innerHeight]);
		};
		window.addEventListener('resize', onResize);
		return () => window.removeEventListener('resize', onResize);
	}, []);

	return (
		<Spin spinning={data === null}>
			<Header />
			<D3 width={windowSize[0]} height={windowSize[1]} />
		</Spin>
	);
};

Visualization.propTypes = {
	startCommit: PropTypes.number.isRequired,
	endCommit: PropTypes.number.isRequired,
	repo: PropTypes.string.isRequired,
	setData: PropTypes.func.isRequired,
	data: PropTypes.object
};

Visualization.defaultProps = {
	data: null
};

export default Visualization;
