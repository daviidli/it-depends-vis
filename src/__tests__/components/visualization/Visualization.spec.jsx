import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { useHistory } from 'react-router';
import axios from 'axios';
import { toast } from 'react-toastify';
import Visualization from '../../../components/visualization/Visualization';

jest.mock('react-router');
jest.mock('axios');
jest.mock('react-toastify');
jest.mock('../../../containers/D3Container', () => () => <div>d3 component</div>);
jest.mock('../../../containers/HeaderContainer', () => () => <div>header</div>);

const defaultProps = {
	startCommit: 0,
	endCommit: 1,
	repo: 'repo1',
	data: null,
	setData: jest.fn()
};

const setup = (props = defaultProps) => {
	const wrapper = shallow(<Visualization {...props} />);
	return { wrapper };
};

const setupMount = (props = defaultProps) => {
	const wrapper = mount(<Visualization {...props} />);
	return { wrapper };
};

describe('Visualization component', () => {
	beforeEach(() => {
		jest.clearAllMocks();
	});

	it('should match snapshot', () => {
		const { wrapper } = setup();
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it('should run effect and fetch data', done => {
		const mockPush = jest.fn();
		useHistory.mockImplementation(() => ({ push: mockPush }));
		axios.get.mockImplementation(jest.fn(() => Promise.resolve({ data: 'some data' })));
		toast.error.mockImplementation(jest.fn());
		setupMount();
		setImmediate(() => {
			expect(defaultProps.setData).toHaveBeenCalledTimes(1);
			expect(defaultProps.setData).toHaveBeenCalledWith('some data');
			expect(toast.error).toHaveBeenCalledTimes(0);
			expect(mockPush).toHaveBeenCalledTimes(0);
			done();
		});
	});

	it('should run effect and toast on error', done => {
		const mockPush = jest.fn();
		useHistory.mockImplementation(() => ({ push: mockPush }));
		axios.get.mockImplementation(jest.fn(() => Promise.reject(new Error('oh no'))));
		toast.error.mockImplementation(jest.fn());
		setupMount();
		setImmediate(() => {
			expect(mockPush).toHaveBeenCalledTimes(1);
			expect(mockPush).toHaveBeenCalledWith('/');
			expect(toast.error).toHaveBeenCalledTimes(1);
			expect(toast.error).toHaveBeenCalledWith('oh no');
			expect(defaultProps.setData).toHaveBeenCalledTimes(0);
			done();
		});
	});

	it('should go back home when repo is not set', done => {
		const mockPush = jest.fn();
		useHistory.mockImplementation(() => ({ push: mockPush }));
		const props = { ...defaultProps };
		props.repo = '';
		setupMount(props);
		setImmediate(() => {
			expect(mockPush).toHaveBeenCalledTimes(1);
			expect(mockPush).toHaveBeenCalledWith('/');
			expect(toast.error).toHaveBeenCalledTimes(0);
			expect(props.setData).toHaveBeenCalledTimes(0);
			done();
		});
	});
});
