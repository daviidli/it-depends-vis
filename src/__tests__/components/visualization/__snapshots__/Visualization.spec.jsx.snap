// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visualization component should match snapshot 1`] = `
<Spin
  size="default"
  spinning={true}
  wrapperClassName=""
>
  <Component />
  <Component
    height={768}
    width={1024}
  />
</Spin>
`;
